catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,379 @@
+    if (children == null) {
+      //  Error while initializing... some records may appear twice.
+      LOGGER.debug("Error getting children for [{}]", entry.getName());
+      return false;
+    }
+    boolean errors = true;
+    for (File child : children) {
+      AsyncFileEntry childEntry = new AsyncFileEntry(entry, child);
+      entry.addChild(childEntry);
+      errors = errors && initHelper(childEntry);
+    }
+    return errors;
+  }
+
+  /**
+   * Initializes the object state of the Observer.
+   *
+   * @return true on success, false if an IO error occurs
+   */
+  public boolean initialize() {
+    return initHelper(rootFile);
+  }
+
+  public void destroy() {
+    rootFile.destroy();
+  }
+
+  @VisibleForTesting
+  protected AsyncFileEntry getRootFile() {
+    return rootFile;
+  }
+
+  public void addListener(final AsyncFileAlterationListener listener) {
+    synchronized (listenerLock) {
+      if (this.listener != null) {
+        throw new IllegalStateException("Cannot be two listeners");
+      }
+      this.listener = listener;
+    }
+  }
+
+  public void removeListener() {
+    //  You cannot remove the listener until the checkAndNotify method is done using it.
+    synchronized (listenerLock) {
+      if (listener != null) {
+        this.listener = null;
+      } else {
+        throw new IllegalStateException("Cannot remove a listener from an empty list");
+      }
+    }
+  }
+
+  /**
+   * Fire directory/file created events to the registered listeners.
+   *
+   * @param entry The file entry
+   */
+  private void doCreate(AsyncFileEntry entry, final AsyncFileAlterationListener listener) {
+
+    if (!addToProcessors(entry)) {
+      return;
+    }
+
+    AsyncFileEntry temp = entry.getFromParent();
+
+    if (temp != null) {
+      //  Someone already committed us. so we're done.
+      removeFromProcessors(entry);
+      return;
+    }
+
+    if (!entry.getFile().isDirectory()) {
+
+      LOGGER.trace("Sending create Request for {}", entry.getName());
+
+      listener.onFileCreate(
+          entry.getFile(), new CompletionSynchronization(entry, this::commitCreate));
+    } else {
+      // Directories are always committed and added to the parent IF they
+      // don't already exist
+
+      File[] children = listFiles(entry.getFile());
+      for (File child : children) {
+        doCreate(new AsyncFileEntry(entry, child), listener);
+      }
+
+      commitCreate(entry, true);
+    }
+  }
+
+  /**
+   * Callback to allow successful ContentFiles to commit themselves
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitCreate(AsyncFileEntry entry, boolean success) {
+
+    LOGGER.debug("commitCreate({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.commit();
+      entry.getParent().ifPresent(e -> e.addChild(entry));
```

Comment:
While there are no sync blocks around this, the entry is still within the processing list so I don't think it will cause any issues. I could be wrong though